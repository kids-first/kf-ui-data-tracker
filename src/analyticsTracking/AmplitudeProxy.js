import {Amplitude} from '@amplitude/react-amplitude';
import {memoize} from '@amplitude/react-amplitude/src/lib/memoize';
import debounce from 'lodash.debounce';
import {buttonTracking} from './eventUtils';
import {EVENT_CONSTANTS} from '../analyticsTracking';

/**
 *
 */
class AmplitudeProxy extends Amplitude {
  logToConsole = false;

  constructor(props) {
    super(props);
    if (typeof props.debounceInterval === 'number') {
      this.logEvent = debounce(this.dispatch, props.debounceInterval);
    } else {
      this.logEvent = this.dispatch;
    }

    this._renderPropParams = {
      getInstance: this.getAmplitudeInstance,
      logEvent: this.logEvent,
      instrument: this.instrument,
      EVENT_CONSTANTS,
      buttonTracking: buttonTracking(this.logEvent),
    };
  }

  // proxy our logging calls so we can hook
  // other analytics services into it
  dispatch = (eventType, eventProps, cb) => {
    if (this.logToConsole || this.props.logToConsole) {
      console.log(
        `AmplitudeProxy::dispatch eventType:${eventType}`,
        eventProps,
      );
    }

    const combinedEventProps = {
      ...this.getAmplitudeEventProperties(),
      ...(eventProps || {}),
    };

    if (
      process.env.NODE_ENV === 'test' ||
      process.env.NODE_ENV === 'CI' ||
      this.props.storeSessionEvents
    ) {
      /** log all events to sessionStorage */
      sessionStorage.setItem(
        'session_events',
        JSON.stringify([
          ...(JSON.parse(sessionStorage.getItem('session_events')) || []),
          {
            eventType,

            eventProps: combinedEventProps,
            utc_time: Date.now(),
          },
        ]),
      );
    }

    return this._makeLogEvent()(eventType, combinedEventProps, cb);
  };

  instrument = memoize((eventType, func, props = {}) => {
    return (...params) => {
      const retVal = typeof func === 'function' ? func(...params) : undefined;
      this.dispatch(eventType, props);

      return retVal;
    };
  });
}

export default AmplitudeProxy;
