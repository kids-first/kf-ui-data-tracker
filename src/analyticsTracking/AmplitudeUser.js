import amplitude from 'amplitude-js';
import jwtDecode from 'jwt-decode';

/**
 * convenience class to log user properties on Auth0 login success
 * @see https://help.amplitude.com/hc/en-us/articles/115002380567-User-Properties-Event-Properties#applying-user-properties-to-events
 */
class AmplitudeUser {
  instance;
  userId;
  auth_sub_arr;

  // decoded jwt values
  auth_user;

  // values to pluck from jwt in the format of [[jwt_prop_name, renmae_to], ... ]
  user_props = [
    ['https://kidsfirstdrc.org/roles', 'roles'],
    ['https://kidsfirstdrc.org/groups', 'studies'],
    ['https://kidsfirstdrc.org/permissions', 'permissions'],
  ];

  constructor(idToken, api_key) {
    amplitude.getInstance().init(api_key);
    this.instance = amplitude.getInstance();
    this.auth_user = jwtDecode(idToken);

    this.auth_sub_arr = this.auth_user.sub.split('|');

    this.setId();
    this.setUserProperies();
    return this;
  }

  /**
   * UUID for amplitude using the parsed value of the
   * auth provider sub field
   * @param {string} id
   */
  setId(id) {
    this.userId = this.auth_sub_arr[1];
    this.instance.setUserId(this.userId);
  }

  setUserProperies() {
    var identify = new this.instance.Identify();

    // pluck properties from dceoded jwt and rename
    this.user_props.forEach(prop => {
      identify.set(prop[1] || prop[0], this.auth_user[prop[0]]);
    });
    // send to amplitude
    this.instance.identify(identify);
  }
}

export default AmplitudeUser;
