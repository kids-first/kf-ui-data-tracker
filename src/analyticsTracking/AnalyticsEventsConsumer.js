import React from 'react';
import {Amplitude} from '@amplitude/react-amplitude';
import useAnalyticsTracking from './useAnalyticsTracking';

const AnalyticsEventsConsumer = ({children, eventProps}) => {
  const analyticsConstants = useAnalyticsTracking();

  return (
    <Amplitude
      eventProperties={inheritedProps => ({
        ...inheritedProps,
        ...eventProps,
      })}
    >
      {props => {
        const buttonTrackingProps = buttonName => ({
          onClick: e =>
            props.logEvent(analyticsConstants.MOUSE.CLICK, {
              button_name: buttonName,
            }),
          onMouseEnter: e =>
            props.logEvent(analyticsConstants.MOUSE.HOVER, {
              button_name: buttonName,
            }),
        });
        if (typeof children === 'function') {
          return children({...props, buttonTrackingProps, analyticsConstants});
        } else {
          return children || null;
        }
      }}
    </Amplitude>
  );
};

export default AnalyticsEventsConsumer;
