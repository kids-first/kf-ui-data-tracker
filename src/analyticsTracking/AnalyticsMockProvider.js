import React from 'react';
import AnalyticsTrackingProvider from './AnalyticsTrackingProvider';

import {Amplitude, LogOnMount} from '@amplitude/react-amplitude';

const AnaltyticsMockProvider = ({children}) => (
  <AnalyticsTrackingProvider>
    <Amplitude eventProperties={{scope: ['TEST']}}>
      <LogOnMount eventType={'TEST_MOUNT'} />
      {children}
    </Amplitude>
  </AnalyticsTrackingProvider>
);

export default AnaltyticsMockProvider;
