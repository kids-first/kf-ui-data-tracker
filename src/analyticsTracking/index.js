export {default as useAnalyticsTracking} from './useAnalyticsTracking';
export {default as withAnalyticsTracking} from './withAnalyticsTracking';

export {
  default as AnalyticsTrackingProvider,
} from './AnalyticsTrackingProvider';

export {default as AmplitudeProxy} from './AmplitudeProxy';
export {default as EVENT_CONSTANTS} from '../common/analyticsTrackingConstants';
