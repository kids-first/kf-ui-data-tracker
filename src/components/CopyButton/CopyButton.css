.CopyButton {
  @extend .inline-block, .relative, .text-xs;
}

.CopyTooltip {
  top: 150%;
  left: 50%;
  border-radius: 4px;
  @extend .absolute, .w-16, .z-10, .-ml-8, .bg-black, .text-white, .text-xs,
    .text-center, .rounded-sm, .py-8;
}

.CopyTooltip:after {
  content: '';
  bottom: 100%;
  left: 50%;
  margin-left: -5px;
  border-width: 5px;
  border-color: transparent transparent black transparent;
  @extend .absolute, .border-solid;
}
