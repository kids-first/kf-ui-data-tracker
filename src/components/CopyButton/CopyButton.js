import React, {useState} from 'react';
import PropTypes from 'prop-types';
import {Button, Icon, Popup} from 'semantic-ui-react';
import {CopyToClipboard} from 'react-copy-to-clipboard';

/**
 * A simple button that is wrapped with an icon and a popup prompting the user
 * to copy the provided text.
 * Passes additional props to the Button
 */
const CopyButton = ({text, ...props}) => {
  const [copied, setCopied] = useState(false);
  return (
    <Popup
      inverted
      position="top left"
      size="mini"
      className='CopyButton--popup'
      trigger={
        <CopyToClipboard
          text={text}
          onCopy={() => {
            setCopied(true);
            setTimeout(() => {
              setCopied(false);
            }, 700);
          }}
        >
          <Button
            {...props}
            icon
            labelPosition="left"
            onClick={e => e.stopPropagation()}
          >
            <Icon name="copy" />
            {text}
          </Button>
        </CopyToClipboard>
      }
    >
      {copied ? <Icon color="green" name="check" /> : 'Copy'}
    </Popup>
  );
};

CopyButton.propTypes = {
  /** The string to be copied */
  text: PropTypes.string.isRequired,
};

CopyButton.defaultProps = {
  text: null,
};

export default CopyButton;
