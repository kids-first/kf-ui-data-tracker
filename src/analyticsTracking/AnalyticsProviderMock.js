import React from 'react';
import AnalyticsTrackingProvider from './AnalyticsTrackingProvider';
import AmplitudeProxy from './services/Amplitude/AmplitudeProxy';

const AnaltyticsMockProvider = ({children}) => (
  <AnalyticsTrackingProvider>
    <AmplitudeProxy eventProperties={{scope: ['TEST']}}>
      {children}
    </AmplitudeProxy>
  </AnalyticsTrackingProvider>
);

export default AnaltyticsMockProvider;
