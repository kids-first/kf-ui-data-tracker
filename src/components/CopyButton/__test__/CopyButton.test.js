import React from 'react';
import {render, cleanup, fireEvent, act} from 'react-testing-library';
import {mocks} from '../../../../__mocks__/kf-api-study-creator/mocks';
import {MockedProvider} from 'react-apollo/test-utils';
import CopyButton from '../CopyButton';
import {AnalyticsProviderMock} from '../../../analyticsTracking';

/* Sets up some browser functions that are needed for react-copy-to-clipboard
 * that do not exist in the test renderer
 */
beforeAll(() => {
  // Always confirm prompts
  global.prompt = () => true;
  // Print alerts to console
  window.alert = () => true;

  // Needed to allow react-copy-to-clipboard to work in test renderer
  const getSelection = () => ({
    rangeCount: 0,
    addRange: () => {},
    getRangeAt: () => {},
    removeAllRanges: () => {},
  });

  window.getSelection = getSelection;
  document.getSelection = getSelection;
});

afterEach(cleanup);

it('Copy Button renders correctly -- default style', async () => {
  const tree = render(
    <AnalyticsProviderMock>
      <MockedProvider mocks={mocks}>
        <CopyButton text="testing1" />
      </MockedProvider>
    </AnalyticsProviderMock>,
  );
  expect(tree.container).toMatchSnapshot();

  // Click on the copy button
  act(() => {
    fireEvent.click(tree.getByText(/testing1/));
  });
});

it('Copy Button renders correctly -- customized style', () => {
  const {container} = render(
    <AnalyticsProviderMock>
      <CopyButton text="testing2" primary basic size="large" />,
    </AnalyticsProviderMock>,
  );
  expect(container).toMatchSnapshot();
});

it('Copy Button renders correctly -- customized tooltip', () => {
  const {container} = render(
    <AnalyticsProviderMock>
      <CopyButton
        text="testing3"
        tooltip="tooltip text"
        position="bottom center"
      />
    </AnalyticsProviderMock>,
  );
  expect(container).toMatchSnapshot();
});

it('Copy Button renders correctly -- customized button icon and text', () => {
  const {container} = render(
    <AnalyticsProviderMock>
      <CopyButton
        text="COPY BUTTON"
        textToCopy="testing4"
        icon="star outline"
      />
    </AnalyticsProviderMock>,
  );
  expect(container).toMatchSnapshot();
});
