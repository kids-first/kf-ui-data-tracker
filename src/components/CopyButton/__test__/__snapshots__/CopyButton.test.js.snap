// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Copy Button renders correctly -- customized button icon and text 1`] = `
<div>
  <button
    class="ui icon button"
  >
    <i
      aria-hidden="true"
      class="star outline icon"
    />
    COPY BUTTON
  </button>
</div>
`;

exports[`Copy Button renders correctly -- customized style 1`] = `
<div>
  <button
    class="ui large basic icon primary button"
  >
    <i
      aria-hidden="true"
      class="copy icon"
    />
    testing2
  </button>
</div>
`;

exports[`Copy Button renders correctly -- customized tooltip 1`] = `
<div>
  <button
    class="ui icon button"
  >
    <i
      aria-hidden="true"
      class="copy icon"
    />
    testing3
  </button>
</div>
`;

exports[`Copy Button renders correctly -- default style 1`] = `
<div>
  <button
    class="ui icon button"
  >
    <i
      aria-hidden="true"
      class="copy icon"
    />
    testing1
  </button>
</div>
`;
